apps/frontend/core-dapp/src/stores/BalancerStore.ts

```diff
@@ -0,0 +1,131 @@
+
+  constructor(root: RootStore) {
+    this.root = root
+
+    makeAutoObservable(this)
+
+    reaction(
+      () => root.web3Store.blockNumber,
+      async (blockNumber) => {
+        const { wstEthPriceInEthBN } = this
+
+        if (
+          wstEthPriceInEthBN !== undefined &&
+          blockNumber !== undefined &&
+          // Only query every 90 blocks (~30s)
+          blockNumber % 90 !== 0
+        )
+          return
+
+        const price = await this.getWstEthPrice()
+        if (price !== undefined) {
+          if (wstEthPriceInEthBN !== undefined && price.eq(wstEthPriceInEthBN)) return
+
+          runInAction(() => {
+            this.wstEthPriceInEthBN = price
+          })
+        }
+      }
+    )
+  }
+
+  get wstEthPriceInEth(): string | undefined {
+    const { wstEthPriceInEthBN } = this
+    if (wstEthPriceInEthBN === undefined) return undefined
+    return utils.formatUnits(wstEthPriceInEthBN, WETH_DECIMALS)
+  }
+
+  private get balancerQueriesAddress(): string | undefined {
+    return getContractAddress<SupportedContracts>(
+      'BALANCER_QUOTER',
+      this.root.web3Store.network.name,
+      supportedContracts
+    )
+  }
+
+  private get wstEthAddress(): string | undefined {
+    const { address } = this.root.baseTokenStore
+    return address
+  }
+
+  private get wethAddress(): string | undefined {
+    const { network } = this.root.web3Store
+    return wethAddressByNetowrk[network.name]
+  }
+
+  private get wstEthWethBalancerPoolId(): string | undefined {
+    const { network } = this.root.web3Store
+    return wstEthWethBalancerPoolIdByNetwork[network.name]
+  }
+
+  private async getWstEthPrice(): Promise<BigNumber | undefined> {
+    const {
+      balancerQueriesInterface,
+      balancerQueriesAddress,
+      wethAddress,
+      wstEthAddress,
+      wstEthWethBalancerPoolId,
+    } = this
+    const { coreProvider } = this.root.web3Store
+
+    if (
+      balancerQueriesAddress === undefined ||
+      wethAddress === undefined ||
+      wstEthAddress === undefined ||
+      wstEthWethBalancerPoolId === undefined
+    )
+      return undefined
+
+    const data = balancerQueriesInterface.encodeFunctionData('querySwap', [
+      {
+        amount: BigNumber.from(10).pow(18),
+        assetIn: wstEthAddress,
+        assetOut: wethAddress,
+        kind: 1, // Given in
+        poolId: wstEthWethBalancerPoolId,
+        userData: [],
+      },
+      {
+        sender: '0x0000000000000000000000000000000000000000',
+        recipient: '0x0000000000000000000000000000000000000000',
+        toInternalBalance: false,
+        fromInternalBalance: false,
+      },
+    ])
+
+    const encodedResponse = await coreProvider.call({
+      to: balancerQueriesAddress,
+      data,
+    })
+
+    const [decodedResponse] = this.balancerQueriesInterface.decodeFunctionResult(
+      'querySwap',
+      encodedResponse
+    )
+
+    return decodedResponse
```

Comment:
Try catch this since call may throw error

```suggestion

    try {
      const encodedResponse = await coreProvider.call({
        to: balancerQueriesAddress,
        data,
      })

      const [decodedResponse] = this.balancerQueriesInterface.decodeFunctionResult(
        'querySwap',
        encodedResponse
      )

      return decodedResponse
    } catch (e) {
      runInAction(() => {
        // cool off for 10secs if rate limited
        this.lastCalled = new Date().getTime() + 10 * SEC_IN_MS
      })
      return this.wstEthPriceInEthBN
    }
```

Reply:
also this way if we get rate limited, at least it will fail gracefully by returning the cached price, and wait for the rate limit to recover

Reply:
> also this way if we get rate limited, at least it will fail gracefully by returning the cached price, and wait for the rate limit to recover

I don't think we want to return the cached price in this function (since it is only called by the reaction to set the price), and in the future we will want to call this function with different wstETH amounts.

Reply:
ahh gotcha, could it return undefined then? I'd like that when rate limited, it doesn't throw error that are uncaught

Reply:
I've added the try/catch [here](https://github.com/prepo-io/prepo-monorepo/pull/957/files#diff-f136e478490ad9b496d51f6959d56d269d288555c8350d40258b7aa1b9e58ee4R46-R55). I don't want to add it on the function because we may want to reuse it somewhere else (i.e. in the withdraw page) and we may want to add our custom error handler there.

Reply:
Also, I'm using a 30s interval which I think it's big enough to not need the extra 10s cooldown